### PR TITLE
Display test results in CodeLens

### PR DESCRIPTION
**Pull request by samcragg**
Hope you don't mind me suggesting an enhancement. Also, it's my first foray into TypeScript so happy to change things to match your preferences. I wanted to display the test result status next to the method, like VS does so I looked at getting the results from the most recent test run.:

preview

The results are outputted to a TRX file and then that file is parsed to get the test results. Finally, the `vscode.executeDocumentSymbolProvider` command is used to get all the symbols in the current document to match the method to the test name.

Things to note. I've tested it with xUnit - it probably needs checking with NUnit as well. Also, I'm using emoji's to display the results, as you can only have text in the CodeLens. This looks fine on Windows 10 but might need a switch to use normal text instead of emoji's in case it looks wrong on other OS's.

Finally, there are things I'd like to enhance but not sure how to implement. I'd like it so you can click on the status and a popup show you the test output (useful for failing tests).

### CODE REVIEW

**Comment by samcragg** [comment]
Sorry about that, still learning the ropes 😄

File: src/testResultsFile.ts [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+    private parseResults(): void {
+        const emitter = this.onNewResultsEmitter;
+        fs.readFile(this.resultsFile, (err, data) => {
+            if (!err) {
+                const results: TestResult[] = [];
+                const xdoc = new DOMParser().parseFromString(data.toString(), "application/xml");
+                const nodes = xdoc.documentElement.getElementsByTagName("UnitTestResult");
+
+                // TSLint wants to use for-of here, but nodes doesn't support it
+                for (let i = 0; i < nodes.length; i++) { // tslint:disable-line
```
Comment (formulahendry) [inline src/testResultsFile.ts]: You could use `for (const node of nodes)`: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/for...of

Comment (samcragg) [reply src/testResultsFile.ts]: I did try that, however, at runtime it throws an exception `TypeError: undefined is not a function` as I don't believe [HTMLCollection](https://developer.mozilla.org/en-US/docs/Web/API/HTMLCollection) (returned by `getElementsByTagName`) has the `iterator` property so it can be used in for-of. May be there's a better XML library I could use to parse the TRX file?

File: src/testResultsFile.ts [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+        const tempFolder = process.platform === "win32" ? process.env.TEMP : "/tmp";
+        this.resultsFile = path.join(tempFolder, TestResultsFile.ResultsFileName);
+
+        // The change event gets called multiple times, so use a one-second
+        // delay before we read anything to avoid doing too much work
+        const me = this;
+        let timeout: NodeJS.Timer;
+        this.watcher = fs.watch(tempFolder, (eventType, fileName) => {
+            if ((fileName === TestResultsFile.ResultsFileName) && (eventType === "change")) {
+                if (!timeout) {
```
Comment (formulahendry) [inline src/testResultsFile.ts]: I think you could use `clearTimeout` here

Comment (samcragg) [reply src/testResultsFile.ts]: I'm not sure that does the same thing, as it a timeout previously established by calling `setTimeout`, however, we're calling it in the callback so there's nothing to cancel. I'm setting it to null so that I can see if I've already triggered a wait or not. I can introduce another flag variable if that would make the code easier to follow?

File: src/testResultsFile.ts (hunk 1 above)
Comment (formulahendry) [reply src/testResultsFile.ts]: In that case, let's just keep this for loop.

File: src/testResultsFile.ts (hunk 2 above)
Comment (formulahendry) [reply src/testResultsFile.ts]: I think you could refer to https://github.com/formulahendry/vscode-docker-explorer/blob/master/src/dockerTreeBase.ts#L21:31 The logic should be similar.
In current implementation, if the change event lasts for 30 seconds continuously (say every 100ms), `parseResults` will be called 30 times.
With the `clearTimeout`, it would be called only once (if the change event is 100ms, or less then 1000ms).

When the test is running, I guess the trx file will be changed continuously. I am not sure whether the TRX will always be a valid XML and could be read normally during the test is running. If it is, I think current implementation is good, since we could have a quick update when test is running.

File: src/testResultsFile.ts [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+                }
+            }
+        });
+    }
+
+    public dispose(): void {
+        try {
+            this.watcher.close();
+            fs.unlinkSync(this.resultsFile);
+        } catch {
```
Comment (formulahendry) [inline src/testResultsFile.ts]: miss `(error)` --> `catch (error)`

File: src/testResultsFile.ts (hunk 2 above)
Comment (samcragg) [reply src/testResultsFile.ts]: Like it, done! :smile:

File: src/testResultsFile.ts (hunk 3 above)
Comment (samcragg) [reply src/testResultsFile.ts]: Added

File: src/dotnetTestExplorer.ts [hunk 4]
```diff
@@ -177,6 +178,22 @@ export class DotnetTestExplorer implements TreeDataProvider<TestNode> {
...
+    private getDotNetTestOptions(): string {
+        return this.checkBuildOption() + this.checkRestoreOption();
+    }
+
+    /**
+     * @description
+     * Gets the dotnet test argument to speicfy the output for the test results.
+     */
+    private outputTestResults(): string {
+        return " --logger 'trx;LogFileName=" + this.resultsFile.fileName + "'";
```
Comment (formulahendry) [inline src/dotnetTestExplorer.ts]: I test it on Windows, it is failed, After I change single quote to double quote, it is working. How about you?

Comment (formulahendry) [reply src/dotnetTestExplorer.ts]: image

File: src/testStatusCodeLensProvider.ts [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+        }
+
+        const results = this.testResults;
+        return commands.executeCommand<SymbolInformation[]>("vscode.executeDocumentSymbolProvider", document.uri)
+            .then((symbols) => {
+                const mapped: CodeLens[] = [];
+                for (const symbol of symbols.filter((x) => x.kind === SymbolKind.Method)) {
+                    const fullName = symbol.containerName + "." + symbol.name;
+                    for (const result of results.values()) {
+                        if (result.testName.endsWith(fullName)) {
```
Comment (formulahendry) [inline src/testStatusCodeLensProvider.ts]: There is a corner case:
image

File: src/dotnetTestExplorer.ts [hunk 6]
```diff
@@ -198,7 +215,7 @@ export class DotnetTestExplorer implements TreeDataProvider<TestNode> {
         return new Promise((c, e) => {
             try {
                 const results = Executor
-                    .execSync(`dotnet test -t -v=q${this.checkBuildOption()}${this.checkRestoreOption()}`, this.testDirectoryPath)
+                    .execSync(`dotnet test -t -v=q${this.getDotNetTestOptions()}`, this.testDirectoryPath)
```
Comment (formulahendry) [inline src/dotnetTestExplorer.ts]: You did not add `${this.outputTestResults()}`, so the test result will not be shown when VS Code starts. Is it by design?

File: src/testResultsFile.ts [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+import { DOMParser } from "xmldom";
+import { TestResult } from "./testResult";
+
+function getAttributeValue(node, name: string): string {
+    const attribute = node.attributes.getNamedItem(name);
+    return (attribute === null) ? null : attribute.nodeValue;
+}
+
+export class TestResultsFile implements Disposable {
+    private static readonly ResultsFileName = "TestExplorerResults.trx";
```
Comment (formulahendry) [inline src/testResultsFile.ts]: What if we have multiple test project opened in different VS Code?

File: src/dotnetTestExplorer.ts (hunk 6 above)
Comment (formulahendry) [reply src/dotnetTestExplorer.ts]: Oh, please ignore. It only lists the tests.

File: src/testStatusCodeLensProvider.ts [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+        while (this.disposables.length) {
+            this.disposables.pop().dispose();
+        }
+    }
+
+    public get onDidChangeCodeLenses(): Event<void> {
+        return this.onDidChangeCodeLensesEmitter.event;
+    }
+
+    public provideCodeLenses(document: TextDocument, token: CancellationToken): CodeLens[] | Thenable<CodeLens[]> {
```
Comment (formulahendry) [inline src/testStatusCodeLensProvider.ts]: Try with xUnit, NUnit and MSTest. Seems only xUnit is showing the test result icon. It would be great if you could also support NUnit and MSTest. 😃

File: src/testStatusCodeLensProvider.ts [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+    private addTestResults(results: TestResult[]) {
+        for (const result of results) {
+            this.testResults.set(result.testName, result);
+        }
+
+        this.onDidChangeCodeLensesEmitter.fire();
+    }
+
+    private checkEnabledOption(): void {
+        this.enabled = workspace.getConfiguration("csharp").get<boolean>("testsCodeLens.enabled", true);
```
Comment (formulahendry) [inline src/testStatusCodeLensProvider.ts]: Not sure it is better to add an extra boolean setting, in case user want to only show the `run test | debug test`. When user click `run test`, our test result icon would not updated. This may confuse users.

File: src/testStatusCodeLensProvider.ts (hunk 8 above)
Comment (samcragg) [reply src/testStatusCodeLensProvider.ts]: This should now be fixed in the latest version, I've tried xUnit, MSTest and NUnit (the later two have the same format)

File: src/testStatusCodeLensProvider.ts (hunk 9 above)
Comment (samcragg) [reply src/testStatusCodeLensProvider.ts]: Added an extra setting `dotnet-test-explorer.showCodeLens`

File: src/testResultsFile.ts (hunk 7 above)
Comment (samcragg) [reply src/testResultsFile.ts]: I'm now using a temporary random folder to store the results, so now works across multiple instances

File: src/testStatusCodeLensProvider.ts (hunk 5 above)
Comment (samcragg) [reply src/testStatusCodeLensProvider.ts]: The symbol provider doesn't return anything for the namespace and the container of the class is empty, so would need to look at perhaps using omni-sharp directly? There's actually an issue with the test discovery with NUnit that only lists the method names, so if two tests exist with the same method name in different classes, NUnit only returns one of them.

Not sure how you want to approach this one? Accept it as a corner case until the extension uses omni-sharp to parse the files to discover the unit test methods?

File: src/dotnetTestExplorer.ts (hunk 4 above)
Comment (samcragg) [reply src/dotnetTestExplorer.ts]: It did work for me, but I've replaced with double quotes and it still works for me so hopefully that should fix it for you

Comment (formulahendry) [reply src/dotnetTestExplorer.ts]: I guess you are using PowerShell. Both are working on PowerShell, while single quote does not work on CMD.

File: src/testStatusCodeLensProvider.ts (hunk 5 above)
Comment (formulahendry) [reply src/testStatusCodeLensProvider.ts]: I think this corner case could be accepted.

Comment (formulahendry) [reply src/testStatusCodeLensProvider.ts]: If you have interest to look at omni-sharp or other approach, it is warmly welcome! You could send PR in the future.

File: src/testResultsFile.ts [hunk 10]
```diff
@@ -0,0 +1,124 @@
...
+
+    public constructor() {
+        const tempFolder = fs.mkdtempSync(path.join(os.tmpdir(), "test-explorer-"));
+        this.resultsFile = path.join(tempFolder, TestResultsFile.ResultsFileName);
+
+        // The change event gets called multiple times, so use a one-second
+        // delay before we read anything to avoid doing too much work
+        const me = this;
+        let changeDelay: NodeJS.Timer;
+        this.watcher = fs.watch(tempFolder, (eventType, fileName) => {
```
Comment (formulahendry) [inline src/testResultsFile.ts]: Only do this when 'showCodeLens' is true

File: src/dotnetTestExplorer.ts [hunk 11]
```diff
@@ -177,6 +178,22 @@ export class DotnetTestExplorer implements TreeDataProvider<TestNode> {
...
+    private getDotNetTestOptions(): string {
+        return this.checkBuildOption() + this.checkRestoreOption();
+    }
+
+    /**
+     * @description
+     * Gets the dotnet test argument to speicfy the output for the test results.
+     */
+    private outputTestResults(): string {
+        return " --logger \"trx;LogFileName=" + this.resultsFile.fileName + "\"";
```
Comment (formulahendry) [inline src/dotnetTestExplorer.ts]: return empty string when 'showCodeLens' is false

File: src/testResultsFile.ts [hunk 12]
```diff
@@ -0,0 +1,124 @@
...
+}
+
+export class TestResultsFile implements Disposable {
+    private static readonly ResultsFileName = "Results.trx";
+    private onNewResultsEmitter = new EventEmitter<TestResult[]>();
+    private resultsFile: string;
+    private watcher: fs.FSWatcher;
+
+    public constructor() {
+        const tempFolder = fs.mkdtempSync(path.join(os.tmpdir(), "test-explorer-"));
```
Comment (formulahendry) [inline src/testResultsFile.ts]: Same here. Only do this when 'showCodeLens' is true. 
Just want to make sure no unnecessary code is running

File: src/testResultsFile.ts [hunk 13]
```diff
@@ -0,0 +1,124 @@
...
+                changeDelay = setTimeout(() => {
+                    me.parseResults();
+                }, 1000);
+            }
+        });
+    }
+
+    public dispose(): void {
+        try {
+            this.watcher.close();
```
Comment (formulahendry) [inline src/testResultsFile.ts]: Also handle when when 'showCodeLens' is false

File: src/testResultsFile.ts (hunk 10 above)
Comment (samcragg) [reply src/testResultsFile.ts]: Good catch, done

File: src/testResultsFile.ts [hunk 14]
```diff
@@ -0,0 +1,134 @@
...
+        });
+    }
+
+    private watchFolder(folder: string): void {
+        // The change event gets called multiple times, so use a one-second
+        // delay before we read anything to avoid doing too much work
+        const me = this;
+        let changeDelay: NodeJS.Timer;
+        this.watcher = fs.watch(folder, (eventType, fileName) => {
+            if ((fileName === TestResultsFile.ResultsFileName) && (eventType === "change")) {
```
Comment (formulahendry) [inline src/testResultsFile.ts]: Test it on macOS, when first run the test, it will only have **eventType === "rename"**. So we could just remove **(eventType === "change")**

File: src/testStatusCodeLens.ts [hunk 15]
```diff
@@ -0,0 +1,25 @@
+"use strict";
+import { CodeLens, Range } from "vscode";
+
+export class TestStatusCodeLens extends CodeLens {
+    public static parseOutcome(outcome: string): string {
+        if (outcome === "Passed") {
+            return "✔️";
```
Comment (formulahendry) [inline src/testStatusCodeLens.ts]: Test on macOS, it is showing a black check mark (https://emojipedia.org/heavy-check-mark/) which is hard to see. 
Two improvements:

- Using another better check mark: https://emojipedia.org/white-heavy-check-mark/. Or you may have a better one? 😄 
- Using the unicode, it would be good to maintain 
image

File: src/testStatusCodeLens.ts [hunk 16]
```diff
@@ -0,0 +1,25 @@
...
+        } else if (outcome === "Failed") {
+            return "❌";
```
Comment (formulahendry) [inline src/testStatusCodeLens.ts]: Same here, using unicode: https://emojipedia.org/cross-mark/

File: src/testStatusCodeLens.ts [hunk 17]
```diff
@@ -0,0 +1,25 @@
...
+        } else if (outcome === "NotExecuted") {
+            return "️️⚠️";
```
Comment (formulahendry) [inline src/testStatusCodeLens.ts]: Same here, using unicode

File: src/testResultsFile.ts (hunk 14 above)
Comment (samcragg) [reply src/testResultsFile.ts]: Removed

File: src/testStatusCodeLens.ts (hunk 15 above)
Comment (samcragg) [reply src/testStatusCodeLens.ts]: I've added settings now so the user can override what we display.

I've kept the heavy check mark for platforms apart from OS X, in which case it uses the white heavy check mark (I'm not a big fan of it, but I can't find any alternatives).

Since the user can specify the text, they could even switch to "passing", "FAILING" etc if they want so hopefully this is OK?

File: src/utility.ts [hunk 18]
```diff
@@ -1,12 +1,49 @@
...
     public static getConfiguration(): vscode.WorkspaceConfiguration {
         return vscode.workspace.getConfiguration("dotnet-test-explorer");
     }
+
+    public static updateCache() {
+        const configuration = Utility.getConfiguration();
+        const osx = platform() === "darwin";
+
+        Utility.showCodeLens = configuration.get<boolean>("showCodeLens", true);
+        Utility.failed = Utility.getLensText(configuration, "codeLensPassed", "\u274c"); // Cross Mark
```
Comment (formulahendry) [inline src/utility.ts]: codeLensPassed?

File: src/utility.ts [hunk 19]
```diff
@@ -1,12 +1,49 @@
...
+        Utility.passed = Utility.getLensText(configuration, "codeLensPassed", osx ? "\u2705" : "\u2714"); // White Heavy Check Mark / Heavy Check Mark
+        Utility.skipped = Utility.getLensText(configuration, "codeLensPassed", "\u26a0"); // Warning
```
Comment (formulahendry) [inline src/utility.ts]: codeLensPassed?

File: src/testStatusCodeLens.ts (hunk 15 above)
Comment (formulahendry) [reply src/testStatusCodeLens.ts]: Looks great to me!

File: src/utility.ts (hunk 18 above)
Comment (samcragg) [reply src/utility.ts]: Good catch, I've actually tested them all this time before submitting (and had my ☕️!)

**Comment by formulahendry** [comment]
@samcragg , it is all good for C# in xUnit, NUnit and MSTest, also Windows and macOS. I just try it on F# but not show the CodeLens. I think [F# extension](https://marketplace.visualstudio.com/items?itemName=Ionide.Ionide-fsharp) also implements the `executeDocumentSymbolProvider` similar to C# extension. Would you please take a look? Or you want to support C# first?

**Comment by formulahendry** [comment]
I just realize that you registerCodeLensProvider for csharp only. I think it is good enough to support C# first. 😃 Thank you for your contribution! This feature is awesome! If you have interest, welcome PR to support F# or any other features.

**Comment by samcragg** [comment]
Thanks! I'll take a look at F# and submit another PR.

I also want to make it so you can click on the lens and it shows you the test output inline (similar to how clicking on the references CodeLens shows them inline) so you can see why a test failed without having to dig through the output yourself (the output is in the XML, just need somewhere to display it).

**Comment by formulahendry** [comment]
Thanks @samcragg ! That would be great!
